src/MongoDB.Driver.Core/Core/Configuration/ConnectionString.cs

```diff
@@ -759,8 +759,17 @@ private void Parse()
                 @"((?<username>[^:@]+)(:(?<password>[^:@]*))?@)?" +
                 serversPattern + @"(/" + databasePattern + ")?/?" + optionsPattern + "$";
 
+            Match badPercentMatch = null;
+            var hexDigit= "0-9A-Fa-f";
+            var percentPattern = $"(%[^{hexDigit}])|"  // % followed by a non-hexadecimal digit OR
+                                 +$"(%[{hexDigit}][^{hexDigit}])"; // % followed by a hexadecimal digit and a non-hexadecimal digit
+            if (_originalConnectionString.Contains("%"))
+            {
+                badPercentMatch = Regex.Match(_originalConnectionString, percentPattern);
+            }
+
             var match = Regex.Match(_originalConnectionString, pattern);
-            if (!match.Success)
+            if (!match.Success || (match.Success && badPercentMatch!=null && badPercentMatch.Success))
```

Comment:
This if statement is complicated. In part because match.Success is in there twice.

If we validate % escape sequences separately as suggested above this line doesn't need to change.